homeassistant/components/connectedcars/config_flow.py

```diff
@@ -0,0 +1,143 @@
+    {
+        vol.Required("namespace", default=DEFAULT_NAMESPACE): str,
+        vol.Required("username"): str,
+        vol.Required("password"): str,
+        vol.Optional("vin"): str,
+    }
+)
+
+
+class ConnectedcarsApiHandler:
+    """Connected Cars.io API connector class."""
+
+    def __init__(self, namespace):
+        """Initialize."""
+        self.namespace = namespace
+        self.vin = None
+        self.client = None
+        self.user_data = None
+
+    async def authenticate(self, username, password) -> bool:
+        """Test if we can authenticate with the namespace."""
+        client = ConnectedCarsClient(username, password, self.namespace)
+
+        self.client = client
+
+        try:
+            self.user_data = await client.async_query(QUERY_USER)
+        except ConnectedCarsException:
+            return False
+
+        return True
+
+    async def get_email(self) -> str:
+        """Will get the email address of the account holder."""
+        if not self.user_data:
+            self.user_data = await self.client.async_query(QUERY_USER)
+
+        email = self.user_data["data"]["viewer"]["email"]
+
+        return email
+
+    async def get_vin(self, userinput_vin) -> str:
+        """Will get the vin identifier of the car."""
+        if not self.vin:
+            response = await self.client.async_query(QUERY_VEHICLE_VIN)
+            # This needs to lookup the "userinput_vin" to see if it can be found in the result, not just take the first car
+            vin = response["data"]["viewer"]["vehicles"][0]["vehicle"]["vin"]
+            self.vin = vin
+
+        return self.vin
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    ccah = ConnectedcarsApiHandler(data["namespace"])
+
+    if not await ccah.authenticate(data["username"], data["password"]):
+        raise InvalidAuth
+
+    try:
+        email = await ccah.get_email()
+    except ConnectedCarsException as exception:
+        raise CannotGetEmail from exception
+
+    userinput_vin = None
+    # if not data["vin"]: ## Do this somehow
+    #     userinput_vin = data["vin"]
+
+    try:
+        vin = await ccah.get_vin(userinput_vin)
+    except ConnectedCarsException as exception:
+        raise CannotGetVin from exception
+
+    integration_title = f"{email} - {vin}"
+
+    # Return info that you want to store in the config entry.
+    return {"title": integration_title}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for ConnectedCars.io."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
```

Comment:
The validate input method isn't used multiple times.

It think the code would become more readable if the logic was just in place here. It removes the need for an additional class and methods, and removes the need for all the additional errors defined.

Reply:
Ok, I will do something about that